apis/python/src/tiledbsoma/_tdb_handles.py

```diff
@@ -244,6 +268,110 @@ def _do_initial_reads(self, reader: tiledb.Group) -> None:
+
+        This is passed a raw TileDB object opened in read mode, since writers
+        will need to retrieve data from the backing store on setup.
+        """
+        # non–attrs-managed field
+        self.metadata = MetadataWrapper(self, dict(reader.meta))
+
+    @property
+    def schema(self) -> pa.Schema:
+        return self._handle.schema
+
+    @property
+    def meta(self) -> "MetadataWrapper":
+        return MetadataWrapper(self, dict(self._handle.meta))
+
+    @property
+    def domain(self) -> Tuple[Tuple[Any, Any], ...]:
+        result = []
+        for name in self._handle.index_column_names:
+            dtype = self._handle.schema.field(name).type
+            if pa.types.is_timestamp(dtype):
+                np_dtype = np.dtype(dtype.to_pandas_dtype())
+                dom = self._handle.domain(name, np_dtype)
+                result.append(
+                    (
+                        np_dtype.type(dom[0], dtype.unit),
+                        np_dtype.type(dom[1], dtype.unit),
+                    )
+                )
+            else:
+                if pa.types.is_large_string(dtype) or pa.types.is_string(dtype):
+                    dtype = np.dtype("U")
+                elif pa.types.is_large_binary(dtype) or pa.types.is_binary(dtype):
+                    dtype = np.dtype("S")
+                else:
+                    dtype = np.dtype(dtype.to_pandas_dtype())
+                result.append(self._handle.domain(name, dtype))
+        return tuple(result)
+
+    @property
+    def ndim(self) -> int:
+        return int(len(self._handle.index_column_names))
+
+    def non_empty_domain(self) -> Optional[Tuple[Tuple[Any, Any], ...]]:
+        result = []
+        for name in self._handle.index_column_names:
+            dtype = self._handle.schema.field(name).type
+            if pa.types.is_timestamp(dtype):
+                np_dtype = np.dtype(dtype.to_pandas_dtype())
+                ned = self._handle.non_empty_domain(name, dtype)
+                result.append(
+                    (
+                        np_dtype.type(ned[0], dtype.unit),
+                        np_dtype.type(ned[1], dtype.unit),
+                    )
+                )
+            else:
+                if pa.types.is_large_string(dtype) or pa.types.is_string(dtype):
+                    dtype = np.dtype("U")
+                elif pa.types.is_large_binary(dtype) or pa.types.is_binary(dtype):
+                    dtype = np.dtype("S")
+                else:
+                    dtype = np.dtype(dtype.to_pandas_dtype())
+                result.append(self._handle.non_empty_domain(name, dtype))
+        return None if len(result) == 0 else tuple(result)
+
+    @property
+    def attr_names(self) -> Tuple[str, ...]:
+        result = []
+        for field in self.schema:
+            if field.name not in self._handle.index_column_names:
+                result.append(field.name)
+        return tuple(result)
```

Comment:
`tuple(f.name for f in self.schema if f.name not in self._handle.index_column_names)`